### PR TITLE
Winforms Tree widget

### PR DESCRIPTION
**Pull request by Oliver-Leigh**
Welcome to the Tree widget for the Windows platform!

### How it works:
The `Tree` widget is based on the `Table` widget, except with some extra structure: 
- The display data is managed by a new class called `StateTree` which is a wrapper for the `TreeSource`. It keeps track of which nodes are expanded/collapsed and then makes/modifies a display list which is sent to the UI
- The nodes are expanded/contracted by clicking a state-change arrow on the left of the row. The state-change arrow also highlights when the cursor is hovering over it. These functionalities are achieved through mouse event actions and a hit-test which determines whether the event has occurred in an appropriate area.
- The non-leaf nodes are custom painted by handling the NM_CUSTOMDRAW message using the subclass method from #4164. 

### Notes on Styling:
The styling is based on that of ListView groups, with some modifications. Unfortunately the combination of virtual mode and groups  is not supported in WinForms, and is undocumented for Win32. Also, looking at the expansion arrow in the picture on the left below, I don't think that this style is being actively updated.
lv_groups_1 lv_groups_2

The solution is to simply custom draw the row. Here is the current iteration of the Tree widget:
widget_1 widget_2

- I've put the state-change arrows on the left to align with the macOS and GTK implementations. 
- The colors used by the native ListView groups are undocumented (as far as I know) so I've used the system color _hotlight_ which is a close match. 
- Since the whole row is custom painted, any of the styling can be changed! 

### Notes for testing:
One of the hardest parts was to get the selection to play nicely with the state-change clicks. The interactions between the clicks-hit-tests and selection is somewhat complicated. Make note of any weirdness you observe in this area! 

### Changes to the Table widget
I've made some minor changes to the WinForms `Table` widget code. This is mainly tidying things up a little, but also I've modified some methods so they can be more easily shared with the `Tree` widget. 

## PR Checklist:

- [x] All new features have been tested
- [ ] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: winforms/src/toga_winforms/widgets/table.py [hunk 1]
```diff
@@ -128,17 +115,81 @@ def _subclass_proc(
...
+        selected_indices = list(self.native.SelectedIndices)
+        if self._multiple_select:
+            return selected_indices
+        elif len(selected_indices) == 0:
+            return None
+        else:
+            return selected_indices[0]
+
+    #################################################################################
+    # The following methods are shared (non-overridden) with  DetailedList and Tree
```
Comment (freakboy3742) [inline winforms/src/toga_winforms/widgets/table.py]: Really minor cosmetic fix
```suggestion
    # The following methods are shared (non-overridden) with DetailedList and Tree
```

File: winforms/src/toga_winforms/widgets/tree.py [hunk 2]
```diff
@@ -0,0 +1,955 @@
...
+            return 0
+        else:
+            return len(self.children)
+
+    def __getitem__(self, row_path: list[int] | tuple[int] | None):
+        """Gets an item based on a row path list or tuple
+
+        :param list[int] | tuple[int] row_path: A list or tuple of indices. The first
+            is an index of a child, say A, in the list of children, the next is an
+            index in the list of children of A, and so on...
```
Comment (freakboy3742) [inline winforms/src/toga_winforms/widgets/tree.py]: There's no need to include the type here; MkDocs will pick it up from the function's type annotation.
```suggestion
        :param row_path: A list or tuple of indices. The first is an index of a child,
            say A, in the list of children, the next is an index in the list of children
            of A, and so on...
```

File: winforms/src/toga_winforms/widgets/tree.py [hunk 3]
```diff
@@ -0,0 +1,955 @@
...
+        notify_select = (
+            self._display_list_modifier(insert, [state_node], index) or notify_select
+        )
+        if not insert:
+            return notify_select
+
+        return False
+
+    def _display_list_modifier(
+        self, insert: bool, sublist: list[StateNode], start_index: int
```
Comment (freakboy3742) [inline winforms/src/toga_winforms/widgets/tree.py]: When argument lists get long enough that we need to split them onto a separate line, we try to go directly to "1 argument per line" format. They're both "ruff accepted" formats, but 1-arg-per-line is a bit easier to read, and makes future argument modifying diffs easier to parse.
```suggestion
        self,
        insert: bool,
        sublist: list[StateNode],
        start_index: int,
```

File: winforms/src/toga_winforms/widgets/tree.py [hunk 4]
```diff
@@ -0,0 +1,955 @@
...
+    ) -> bool:
+        """Modifies the display list by either inserting or removing a sublist.
+
+        The sublist contains either a single element, or a whole branch which will be
+        expanded/contracted during a toggle operation.
+
+        :param bool insert: Is the sublist being inserted or removed?
+        :param int start_index: The index in the display list where the first item
+            of the sublist is/will be.
```
Comment (freakboy3742) [inline winforms/src/toga_winforms/widgets/tree.py]: As noted previously - no need for types here
```suggestion
        :param insert: Is the sublist being inserted or removed?
        :param start_index: The index in the display list where the first item
            of the sublist is/will be.
```

File: winforms/src/toga_winforms/widgets/tree.py [hunk 5]
```diff
@@ -0,0 +1,955 @@
...
+        """
+        for state_node in self.branch_iter(display=False):
+            if state_node.node == node:
+                return state_node
+
+        return None
+
+
+def display_branch_iter(
+    roots: Iterable[StateNode | StateTree], ignore_closed: bool = False
```
Comment (freakboy3742) [inline winforms/src/toga_winforms/widgets/tree.py]: As noted previously - this should be 1-arg-per-line format

File: winforms/src/toga_winforms/widgets/tree.py (hunk 2 above)
Comment (Oliver-Leigh) [reply winforms/src/toga_winforms/widgets/tree.py]: No problem! I've removed them everywhere.

File: winforms/src/toga_winforms/widgets/tree.py (hunk 3 above)
Comment (Oliver-Leigh) [reply winforms/src/toga_winforms/widgets/tree.py]: Fixed!

**Comment by Oliver-Leigh** [comment]
> The app is currently using an empty string for each "non column 1" field. That's probably a change we should make to the example, so that it's obvious - but with that change, in the Windows rendering, as soon as a leaf becomes a node, we lose the text for the rest of the node.

Oh I completely missed that! Also for the example, it might be nice to have some higher depth nodes as well. 


> You've mentioned that the row style is customisable - is that sort of change in the realm of the possible? I'm guessing that might mean we lose the blue bar - I'm not sure if that's a "dealer's choice" thing from you, or if it's idiomatic on Windows.

We can have the extra columns and have blue bars between the non-empty columns (if that's something we want?). I'll have a think about the best way to do it and make the changes!

**Comment by freakboy3742** [comment]
> > The app is currently using an empty string for each "non column 1" field. That's probably a change we should make to the example, so that it's obvious - but with that change, in the Windows rendering, as soon as a leaf becomes a node, we lose the text for the rest of the node.
> 
> Oh I completely missed that! Also for the example, it might be nice to have some higher depth nodes as well.

Agreed - maybe adding a 1900s and 2000s grandparent nodes? If we need to [add some 2010s data to make that separation make sense...](https://www.imdb.com/find/?q=bees)
 
> We can have the extra columns and have blue bars between the non-empty columns (if that's something we want?). I'll have a think about the best way to do it and make the changes!

I guess my question is the extent to which the blue bar is idiomatic. If it is, then yes, retaining it makes sense; if it's not, then I wonder if it's more trouble than it's worth, and we should stick to just the expansion arrow.

**Comment by Oliver-Leigh** [comment]
> Agreed - maybe adding a 1900s and 2000s grandparent nodes? If we need to [add some 2010s data to make that separation make sense...](https://www.imdb.com/find/?q=bees)

Good idea! I can look into it after this is finished. 

> I guess my question is the extent to which the blue bar is idiomatic. If it is, then yes, retaining it makes sense; if it's not, then I wonder if it's more trouble than it's worth, and we should stick to just the expansion arrow.

Good question. I think you'd have a hard time arguing that either the blue or the blue bar is idiomatic of Windows as a whole, simply because the style is always changing. I've decided to remove both of these. I can always put them back in if there's popular demand!

**Comment by Oliver-Leigh** [comment]
One extra thing to note is that I've disabled the focus rectangle for now. Getting it to work properly was becoming very complicated. I'll have to think a bit more about how to do it properly.

**Review by freakboy3742 — APPROVED** [review]
I needed to make a small fix for compatibility with a recent change in how widgets are registered (which was the source of the test failure); I also needed to test the "text in node columns" thing, so I've taken the opportunity to update the tree example to have multiple layers of depth and text in the non-leaf columns.

Otherwise, this is incredible - thanks for filling a really notable gap in the Winforms backend!

**Comment by Oliver-Leigh** [comment]
> I needed to make a small fix for compatibility with a recent change in how widgets are registered (which was the source of the test failure);

Thanks! I thought it would be something like this, but I didn't want to make a mess. 

> Looks like there are some test coverage issues.
> 
> We had a similar issue with the Qt backend - there's some edge cases that need to be caught for error handling completeness purposes, but they either _should_ never happen, or they can't be reproduced reliably under testbed conditions.
> 
> We don't have a problem adding `# pragma: no cover` to the code if there are cases that _can't_ be covered - but they should always be accompanied with a comments backing up why a `no cover` is valid under the circumstances.
> 
> It may also be necessary to add additional test cases; if that's the case, feel free to add extra steps to an existing testbed case, or an entirely new testbed case, as required.

I've fixed the coverage issues now. Most of them were because of the various mouse event handlers that are being used. 

When writing these tests I found an interesting discrepancy between the platforms: On macOS and Windows, when a selected item is hidden, it is deselected. However this is not the case on Linux. I'm not sure what the desired behavior is. I based the behavior of the Windows version on the macOS version, so those two align at least.

File: winforms/src/toga_winforms/widgets/tree.py [hunk 6]
```diff
@@ -693,6 +697,7 @@ def winforms_mouse_leave(self, sender, e):
         movement beginning at the state-change arrow will not register as a
         MouseClick or MouseUp event.
         """
+        print("here")
```
Comment (freakboy3742) [inline winforms/src/toga_winforms/widgets/tree.py]: Stray debug?
```suggestion
```

File: testbed/tests/widgets/test_tree.py [hunk 7]
```diff
@@ -228,6 +224,64 @@ async def test_select(widget, probe, source, on_select_handler):
...
+    assert probe.is_expanded(source[3])
+    await probe.select_row((4,))
+    await probe.redraw(
+        "A node is expanded and a root-row with higher index is selected"
+    )
+    widget.collapse(source[3])
+    assert widget.selection == source[4]
+    assert not probe.is_expanded(source[3])
+
+    if toga.platform.current_platform in {"macOS", "windows"}:
```
Comment (freakboy3742) [inline testbed/tests/widgets/test_tree.py]: I can see what you're doing here, but we don't use this pattern in testbed. Details in the overall review comments.

File: winforms/src/toga_winforms/widgets/tree.py (hunk 6 above)
Comment (Oliver-Leigh) [reply winforms/src/toga_winforms/widgets/tree.py]: Fix!

**Comment by Oliver-Leigh** [comment]
I've cleaned up the tests as you suggested. In the end there is the `test_mouse_events` test which I use pytest.skip for the other platforms, and two instances of exercising but not asserting: 

- Expand/contract a non-visible node.
- Contract a node whose child is selected. 

Also the focused item rectangle has been implemented! It's a bit more complicated than I'd like, but that's due to some shortcomings shared between Win32 and WinForms

**Comment by johnzhou721** [comment]
> One extra thing to note is that I've disabled the focus rectangle for now. Getting it to work properly was becoming very complicated. I'll have to think a bit more about how to do it properly.

For the record:  Focus rectangles on Windows is not idiomatic anymore; on modern WinUI apps there's no distinction between "selected" and "focussed".  So this is actually a good thing overall.
